src/rt/alloc.rs

```diff
@@ -0,0 +1,64 @@
+use crate::rt;
+use crate::rt::object::Object;
+
+/// Tracks an allocation
+#[derive(Debug)]
+pub(crate) struct Allocation {
+    obj: Object,
+}
+
+#[derive(Debug)]
+pub(super) struct State {
+    is_dropped: bool,
+}
+
+/// Track a raw allocation
+pub(crate) fn alloc(ptr: *mut u8) {
+    rt::execution(|execution| {
+        let obj = execution.objects.insert_alloc(State { is_dropped: false });
+
+        let allocation = Allocation { obj };
+
+        let prev = execution.raw_allocations.insert(ptr as usize, allocation);
+        assert!(prev.is_none(), "pointer already tracked");
+    });
+}
+
+/// Track a raw deallocation
+pub(crate) fn dealloc(ptr: *mut u8) {
+    let allocation =
+        rt::execution(
+            |execution| match execution.raw_allocations.remove(&(ptr as usize)) {
+                Some(allocation) => allocation,
+                None => panic!("pointer not tracked"),
+            },
+        );
+
+    // Drop outside of the `rt::execution` block
+    drop(allocation);
+}
+
+impl Allocation {
+    pub(crate) fn new() -> Allocation {
+        rt::execution(|execution| {
+            let obj = execution.objects.insert_alloc(State { is_dropped: false });
+
+            Allocation { obj }
+        })
+    }
+}
+
+impl Drop for Allocation {
+    fn drop(&mut self) {
+        rt::execution(|execution| {
+            let state = self.obj.alloc(&mut execution.objects);
+            state.is_dropped = true;
+        });
+    }
+}
+
+impl State {
+    pub(super) fn check_for_leaks(&self) {
+        assert!(self.is_dropped, "object leaked");
```

Comment:
Since this is an assert, it will panic immediately when the first leaked allocation is checked. It seems like we might (eventually) want to change this to return a `bool` or something, so that `Execution::check_for_leaks` can count the number of leaked `Arc`s and raw allocations, and panic if either are non-zero, so that all leaks that occurred during an execution can be reported.

This would be especially nice if we also started tracking unique metadata (e.g. stack frames) per allocated object.